src/vdaf/mastic.rs

```diff
@@ -0,0 +1,473 @@
+    use super::*;
+    use crate::field::Field128;
+    use crate::flp::types::Sum;
+    use rand::{thread_rng, Rng};
+
+    const TEST_WEIGHT_LEN: usize = 5;
+    const TEST_NONCE_SIZE: usize = 16;
+
+    #[test]
+    fn test_mastic_shard_sum() {
+        let algorithm_id = 6;
+        let sum_typ = Sum::<Field128>::new(5).unwrap();
+        let sum_szk = Szk::new_turboshake128(sum_typ, algorithm_id);
+        let sum_vidpf = Vidpf::<VidpfWeight<Field128>, TEST_NONCE_SIZE>::new(TEST_WEIGHT_LEN);
+
+        let mut nonce = [0u8; 16];
+        let mut verify_key = [0u8; 16];
+        thread_rng().fill(&mut verify_key[..]);
+        thread_rng().fill(&mut nonce[..]);
+
+        let first_input = VidpfInput::from_bytes(&[15u8, 0u8, 1u8, 4u8][..]);
+
+        let mastic = Mastic::new(algorithm_id, sum_szk, sum_vidpf, 32);
+        let (_public, _input_shares) = mastic.shard(&(first_input, 24u128), &nonce).unwrap();
+    }
+
+    #[test]
+    fn test_input_share_encode() {
+        let algorithm_id = 6;
+        let sum_typ = Sum::<Field128>::new(5).unwrap();
+        let sum_szk = Szk::new_turboshake128(sum_typ, algorithm_id);
+        let sum_vidpf = Vidpf::<VidpfWeight<Field128>, TEST_NONCE_SIZE>::new(TEST_WEIGHT_LEN);
+
+        let mut nonce = [0u8; 16];
+        let mut verify_key = [0u8; 16];
+        thread_rng().fill(&mut verify_key[..]);
+        thread_rng().fill(&mut nonce[..]);
+
+        let first_input = VidpfInput::from_bytes(&[15u8, 0u8, 1u8, 4u8][..]);
+
+        let mastic = Mastic::new(algorithm_id, sum_szk, sum_vidpf, 32);
+        let (_public, input_shares) = mastic.shard(&(first_input, 26u128), &nonce).unwrap();
+        let leader_input_share = &input_shares[0];
+        let helper_input_share = &input_shares[1];
+
+        assert_eq!(
+            leader_input_share.encoded_len().unwrap(),
+            leader_input_share.get_encoded().unwrap().len()
+        );
+        assert_eq!(
+            helper_input_share.encoded_len().unwrap(),
+            helper_input_share.get_encoded().unwrap().len()
+        );
+    }
+
+    #[test]
+    fn test_input_share_roundtrip() {
+        let algorithm_id = 6;
+        let sum_typ = Sum::<Field128>::new(5).unwrap();
+        let sum_szk = Szk::new_turboshake128(sum_typ, algorithm_id);
+        let sum_vidpf = Vidpf::<VidpfWeight<Field128>, TEST_NONCE_SIZE>::new(TEST_WEIGHT_LEN);
+
+        let mut nonce = [0u8; 16];
+        let mut verify_key = [0u8; 16];
+        thread_rng().fill(&mut verify_key[..]);
+        thread_rng().fill(&mut nonce[..]);
+
+        let first_input = VidpfInput::from_bytes(&[15u8, 0u8, 1u8, 4u8][..]);
+
+        let mastic = Mastic::new(algorithm_id, sum_szk, sum_vidpf, 32);
+        let (_public, input_shares) = mastic.shard(&(first_input, 25u128), &nonce).unwrap();
+        let leader_input_share = &input_shares[0];
+        let helper_input_share = &input_shares[1];
+
+        let mut bytes = vec![];
+        leader_input_share.encode(&mut bytes).unwrap();
```

Comment:
As with the other encoding tests, `get_encoded()` and `get_decoded()` would be more concise and a more strict test, respectively.